### PR TITLE
Batch persist

### PR DESCRIPTION
**Pull request by HammadB**
This PR adds a batch persist functionality via the persistDirty() nethid to the graph which only persists dirtied elements in the graph. We store data in four files, a header, the data_level_0, the length, and the link lists. The latter three files map to the in memory representation.  Data is never read from disk except on load, serving as a write-through cache. Callers are expected to periodically call persistDirty() in a thread-compatible way.   

This storage scheme is extremely naive, and is only meant as an improvement to serializing the whole index. We can make many improvements in terms of disk access, layout, caching, and durability.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,7 +9,8 @@ if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
     set(CMAKE_CXX_STANDARD 11)
 
     if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-      SET( CMAKE_CXX_FLAGS  "-Ofast -DNDEBUG -std=c++11 -DHAVE_CXX0X -openmp -march=native -fpic -ftree-vectorize")
+      # SET( CMAKE_CXX_FLAGS  "-Ofast -DNDEBUG -std=c++11 -DHAVE_CXX0X -openmp -fpic -ftree-vectorize")
```
Comment (HammadB) [inline CMakeLists.txt:12]: revert this

File: hnswlib/hnswalg.h [hunk 2]
```diff
@@ -72,6 +73,12 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
     std::unordered_set<tableint> deleted_elements;  // contains internal ids of deleted elements
 
 
+    bool persist_on_write_ = false;
+    std::string persist_location_;
+    std::mutex elements_to_persist_lock_; // lock for elements_to_persist_
+    std::set<tableint> elements_to_persist_; // dirty elements to persist
+    int sync_threshold_ = 100; // sync counter, we will sync every sync_threshold_ insertions
```
Comment (HammadB) [inline hnswlib/hnswalg.h:80]: remove this

File: hnswlib/hnswalg.h [hunk 3]
```diff
@@ -656,6 +700,291 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+            return;
+        }
+
+        if (!persist_on_write_){
+            throw std::runtime_error("persistDirty called for an index that is not set to persist on write");
+        }
+        
+        persistHeader(this->output_header_);
+
+        // Note: We could benifet a lot from async IO here. Either via classic POSIX AIO or via libaio
```
Comment (HammadB) [inline hnswlib/hnswalg.h:837]: typo benefit

File: hnswlib/hnswalg.h [hunk 4]
```diff
@@ -656,6 +700,291 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+        output_data_level0.flush();
+
+        // Write lengths
+        output_length.seekp(0, std::ios::beg);
+        output_length.write(length_memory_, max_elements_ * sizeof(float));
+        output_length.flush();
+
+        // Write linklists
+        output_link_lists.seekp(0, std::ios::beg);
+        for (size_t i = 0; i < cur_element_count; i++) {
```
Comment (HammadB) [inline hnswlib/hnswalg.h:782]: cur_element_count is 0, can remove this dead code

File: hnswlib/hnswalg.h [hunk 5]
```diff
@@ -656,6 +700,291 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+
+        // Read data_level0_memory_
+        std::ifstream input_data_level0(this->getDataLevel0Location(), std::ios::binary);
+        if (!input_data_level0.is_open())
+            throw std::runtime_error("Cannot open data_level0 file");
+
+        data_level0_memory_ = (char *) malloc(max_elements * size_data_per_element_);
+        if (data_level0_memory_ == nullptr)
+            throw std::runtime_error("Not enough memory: loadPersistedIndex failed to allocate level0");
+        // TODO: the following reads don't need to be max_elements long
```
Comment (HammadB) [inline hnswlib/hnswalg.h:925]: Remove this todo

File: hnswlib/hnswalg.h [hunk 6]
```diff
@@ -656,6 +700,291 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+        std::vector<std::mutex>(MAX_LABEL_OPERATION_LOCKS).swap(label_op_locks_);
+
+        visited_list_pool_ = new VisitedListPool(1, max_elements);
+
+        // Read the linkLists
+        std::ifstream input_link_list(this->getLinkListLocation(), std::ios::binary);
+        if (!input_link_list.is_open())
+            throw std::runtime_error("Cannot open link list file");
+
+        linkLists_ = (char **) malloc(sizeof(void *) * max_elements);
```
Comment (HammadB) [inline hnswlib/hnswalg.h:954]: Lets pull this out and share it with loadIndex()

File: hnswlib/hnswalg.h [hunk 7]
```diff
@@ -656,6 +700,291 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+        std::ifstream input_header(this->getHeaderLocation(), std::ios::binary);
+        if (!input_header.is_open())
+            throw std::runtime_error("Cannot open header file");
+
+        // Read the header
+        int persisted_version;
+        readBinaryPOD(input_header, persisted_version);
+        // For now, version is a simple equality check, we may add backwards compatibility later
+        if (persisted_version != PERSISTENCE_VERSION)
+            throw std::runtime_error("Cannot read persisted index: wrong persistence version");
```
Comment (HammadB) [inline hnswlib/hnswalg.h:889]: output the persistence version

**Review by atroyn — APPROVED** [review]
LGTM

File: hnswlib/hnswalg.h (hunk 7 above)
Comment (HammadB) [reply hnswlib/hnswalg.h:879]: using std::to_string() is breaking windows due to what seems like a compiler that doesn't support c++11. I don't want to introduce an extra variable so we can revisit this in the future.

**Pull request merged**